File: common/utils/packet_id_generator.py
import threading
import random

class PacketIDGenerator12Bit:
    """12ビット幅のパケットIDを生成するスレッドセーフなジェネレーター"""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._current = random.randint(0, 4095)
        self._max_id = 4096  # 2^12

    def next_id(self) -> int:
        """次のIDを返す"""
        with self._lock:
            pid = self._current
            self._current = (self._current + 1) % self._max_id
            return pid

    def next_id_bytes(self) -> bytes:
        """2バイトに12ビット分を格納して返す"""
        pid = self.next_id()
        return pid.to_bytes(2, byteorder='little')

File: cpp/common/clients/utils/PacketIDGenerator.cpp
#include "PacketIDGenerator.hpp"
#include <random>

namespace wip {
namespace clients {
namespace utils {

PacketIDGenerator::PacketIDGenerator() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<uint16_t> dist(0, MAX_ID - 1);
    current_ = dist(gen);
}

PacketIDGenerator& PacketIDGenerator::instance() {
    static PacketIDGenerator inst;
    return inst;
}

uint16_t PacketIDGenerator::next_id() {
    std::lock_guard<std::mutex> lock(mtx_);
    uint16_t pid = current_;
    current_ = static_cast<uint16_t>((current_ + 1) % MAX_ID);
    return pid;
}

std::array<uint8_t,2> PacketIDGenerator::next_id_bytes() {
    uint16_t pid = next_id();
    std::array<uint8_t,2> bytes{};
    bytes[0] = static_cast<uint8_t>(pid & 0xFF);
    bytes[1] = static_cast<uint8_t>((pid >> 8) & 0x0F);
    return bytes;
}

} // namespace utils
} // namespace clients
} // namespace wip
